core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala

```diff
@@ -203,36 +222,49 @@ class LoadBalancerService(config: WhiskConfig, verbosity: LogLevel)(
     }
 
     /**
-     * Creates an entry in the activation map where we note the time of publishing and establishes a place to store the result.
+     * Creates an activation entry and insert into various maps.
      */
-    private def setupActivation(activationId: ActivationId, timeout: FiniteDuration, transid: TransactionId): ActivationEntry = {
+    private def setupActivation(activationId: ActivationId, subject: String, invokerIndex: Int, timeout: FiniteDuration, transid: TransactionId): ActivationEntry = {
         // either create a new promise or reuse a previous one for this activation if it exists
-        activationMap.getOrElseUpdate(activationId, {
+        val entry = activationById.getOrElseUpdate(activationId, {
             val promise = Promise[WhiskActivation]
-            // store the promise to complete on success, or fail with ActiveAckTimeout
-            // if the time runs out
+            // store the promise to complete on success, or fail with ActiveAckTimeout if time is up
+            // however, do not remove the entry as this is done only by processCompletion
             actorSystem.scheduler.scheduleOnce(timeout) {
-                activationMap.remove(activationId).foreach { _ =>
-                    val failedit = promise.tryFailure(new ActiveAckTimeout(activationId))
-                    if (failedit) info(this, "active response timed out")
+                activationById.get(activationId).foreach { _ =>
+                    if (promise.tryFailure(new ActiveAckTimeout(activationId))) {
+                        info(this, "active response timed out")
+                    }
                 }
             }
-
-            ActivationEntry(Instant.now(Clock.systemUTC()), promise)
+            ActivationEntry(activationId, subject, invokerIndex, Instant.now(Clock.systemUTC()), promise)
         })
+        activationByInvoker.getOrElseUpdate(invokerIndex, new TrieSet[ActivationEntry]).put(entry, {})
+        activationBySubject.getOrElseUpdate(subject, new TrieSet[ActivationEntry]).put(entry, {})
+        entry
+    }
+
+    /**
+      * When invoker health detects a new invoker has come up, this callback is called.
+      */
+    private def invokerChangeCallback(invokerIndices: Array[Int]) = {
+        invokerIndices.foreach { index =>
+            invokerActivationCounter(index) = new Counter()
+            val actSet = activationByInvoker.getOrElseUpdate(index, new TrieSet[ActivationEntry])
+            actSet.keySet map { case ActivationEntry(activationId, subject, invokerIndex, _, promise) =>
+                promise.tryFailure(new LoadBalancerException(s"Invoker $invokerIndex restarted"))
+                activationById.remove(activationId)
+                activationBySubject.remove(subject)
```

Comment:
Is this supposed to kill all the activations from the subject? Shouldn't this just remove the current `ActivationEntry` from the subject's set?